ngraph/core/src/op/proposal.cpp

```diff
@@ -37,51 +37,78 @@ op::v0::Proposal::Proposal(const Output<Node>& class_probs,
 void op::v0::Proposal::validate_and_infer_types()
 {
     NGRAPH_OP_SCOPE(v0_Proposal_validate_and_infer_types);
-    const auto& class_probs_pshape = get_input_partial_shape(0);
-    const auto& class_bbox_deltas_pshape = get_input_partial_shape(1);
-    const auto& image_shape_pshape = get_input_partial_shape(2);
-    if (class_probs_pshape.is_static() && class_bbox_deltas_pshape.is_static() &&
-        image_shape_pshape.is_static())
-    {
-        const Shape class_probs_shape{class_probs_pshape.to_shape()};
-        const Shape class_bbox_deltas_shape{class_bbox_deltas_pshape.to_shape()};
-        const Shape image_shape_shape{image_shape_pshape.to_shape()};
+    const auto& class_probs_ps = get_input_partial_shape(0);
+    const auto& bbox_deltas_ps = get_input_partial_shape(1);
+    const auto& image_shape_ps = get_input_partial_shape(2);
+    auto out_dimension = Dimension::dynamic();
 
+    if (class_probs_ps.is_static())
+    {
         NODE_VALIDATION_CHECK(
             this,
-            class_probs_shape.size() == 4,
+            class_probs_ps.rank().get_length() == 4,
             "Proposal layer shape class_probs input must have rank 4 (class_probs_shape: ",
-            class_probs_shape,
+            class_probs_ps,
             ").");
+    }
 
+    if (bbox_deltas_ps.is_static())
+    {
         NODE_VALIDATION_CHECK(this,
```

Comment:
Same as for `class_probs_ps`. Using rank().compatible(4) could be moved outside if condition